Implement RTS State Machine

Ca1ebK opened the pull request:
Implements a state machine for DUNE FNAL RTS chip testing automation with simulation and hardware modes.

## Key Features
- **15-state machine** with normal operations, error handling, and recovery
- **Flexible chip management**: manual entry or automatic full tray population
- **Interactive pause/resume** system with state tracking
- **Hardware integration** via `MoveChipsToSockets()` and `MoveChipsToTray()`
- **Log-driven transitions** for external system control
- **Comprehensive error handling** with bounds checking and validation

## Technical Details
- Dictionary-based chip position tracking with validation
- Session folder creation for test data storage

The state machine supports both simulation mode for development and real hardware integration for production use.

Ca1ebK committed:
Update README.md

Ca1ebK committed:
Basic files

Ca1ebK committed:
Delete GUI.py

Ca1ebK committed:
Add basic states and transitions

Ca1ebK committed:
Pause on the README

Ca1ebK committed:
Update README.md

Ca1ebK committed:
Add error states and transitions

Expanded the RTS state machine with detailed error states for various fault conditions, including server, socket, movement, testing, and database errors. Updated the README to document new error states, error handling, and recovery mechanisms. Introduced error transition logic in RTSStateMachine.py to enable targeted error handling and recovery from operational states.

Ca1ebK committed:
Revise README and polish code

Ca1ebK committed:
Update README.md

Ca1ebK committed:
Update README.md

Ca1ebK committed:
Update README.md

Ca1ebK committed:
Add more transitions and tests

Ca1ebK committed:
Enhance state machine with advanced pause/resume and error recovery

Added advanced pause/resume functionality with precise state restoration using before_pause_cycle() and resume_to_previous(). Expanded error handling with contextual recovery paths and new state callbacks for error and recovery states. Updated documentation and main.py test to demonstrate new features and improved error recovery logic.

Ca1ebK committed:
Fix state transitions

Ca1ebK committed:
Update main.py

Ca1ebK committed:
To Do: User Input

Ca1ebK committed:
Refactor state machine and add interactive pause

Simplified and condensed the README for clarity and quick start. Enhanced RTSStateMachine with a user-interactive pause/resume method and added a test_cycle transition group for stepwise testing. Updated main.py to use the new test_cycle method for demonstration.

Ca1ebK committed:
Add chip tray management and automate full tray testing

Added chip tray tracking and management to RTSStateMachine, including methods to advance, reset, and check tray completion. Implemented functionality to run a full test cycle for each chip and automatically process all 40 positions on a 4×10 tray. Updated main.py to handle tray iteration and revised the README with usage instructions and method documentation.

Ca1ebK committed:
Update RTSStateMachine.py

Remove unnecessary instance variable, add helpful visual comment

Ca1ebK committed:
Clean up code and fix pause functionality

Removed unnecessary comments (will likely re-comment in the future, added a pause menu functionality)

Still todo: fix transition system logic

Ca1ebK committed:
Update README.md

Ca1ebK committed:
Refactor chip position tracking to use dictionary-based approach

- Replaced individual row/col variables with a dictionary containing 'col' and 'row' lists for chip positions

- Updated state machine methods to use index-based position tracking

- Added helper methods: get_current_chip_data() and set_chip_data()

- Updated README to document new structure and usage

Ca1ebK committed:
Add a helpful visualization for chip_positions dictionary

Ca1ebK committed:
Add a couple more tests and fix instructions

Ca1ebK committed:
Integrate RTS hardware control and refactor state machine

Added FNAL_RTS_integration.py and RTS_CFG.py for hardware integration and robot control. Refactored RTSStateMachine to use a dictionary for chip position tracking, clarified state transitions, and improved pause/resume logic. Updated README with recent changes and modified main.py for new state machine usage.

Ca1ebK committed:
Add session folder creation and RTS bypass logic

Introduced automatic creation of a timestamped session folder upon entering the ground state for session-specific data storage. Added a BypassRTS flag to toggle between simulation and real hardware calls for chip movement, and updated main.py to test both modes. Also refactored chip position population and updated documentation with recent changes and new TODOs.

Ca1ebK committed:
Add log file functionality

Created a function to read a log file and update states based on its contents, updated main.py to demonstrate a test

Ca1ebK committed:
Update log-based state transition and documentation

Expanded the README with details on log-based state transitions and recent changes. Fixed a bug in RTSStateMachine where log monitoring would break if the log file does not exist. Updated main.py to use the correct log file name. Added more state transitions to state_log.txt for testing.

Ca1ebK committed:
Update log reading to include keywords

Ca1ebK committed:
Update README

Updates to chip position dictionary

Ca1ebK committed:
Add interactive chip population and tray handling

Introduced interactive user input for chip position population, allowing manual entry with validation or automatic full tray setup. Integrated MoveChipsToTray in the state machine, ensured chip advancement only in run_full_cycle(), and polished documentation and error handling throughout RTSStateMachine. Updated main.py with a test

tcontrer commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+python-statemachine==2.5.0
```
Can you include the other packages we use here?

tcontrer commented on state_log.txt:
```diff
@@ -0,0 +1 @@
+Picked up chip from tray
```
We probably don't want to save the log file to the git hub, can you remove this?

tcontrer commented on RTSStateMachine.py:
```diff
@@ -0,0 +1,588 @@
+        """Get all data for the current chip as a dictionary."""
+        return {key: self.chip_positions[key][self.current_chip_index] for key in self.chip_positions}
+    
+    def set_chip_data(self, index, col=None, row=None):
```
Can you include a definition for index (possibly all inputs) into this function?

tcontrer reviewed (changes requested):
Made a few very minor comments for you to update, but otherwise looks great

Ca1ebK committed:
Create set_chip_data docstring, update requirements, remove log file

Improved the docstring for set_chip_data in RTSStateMachine.py to clarify arguments. Updated requirements.txt to include pandas and colorama dependencies. Modified main.py to call handle_tray() instead of run_full_cycle(). Removed obsolete state_log.txt.
